File: c#/ValidParentheses.cs
using System;
using System.Collections.Generic;

public class Solution
{
	public bool ValidParentheses(String str)
	{
		if(str.Length % 2 != 0) return false;

		Stack<char> stk = new Stack<char>();

		foreach(char character in str)
		{
			if(character == '(' || character == '[' || character == '{') stk.Push(character);

			if(stk.Count == 0)
			{
				if(character == ')' || character == ']' || character == '}') return false;
			}

			if(stk.Count != 0)
			{
				if(character == ')' && stk.Peek() == '(') stk.Pop();
				else if(character == ')' && stk.Peek() != '(') return false;

				if(character == ']' && stk.Peek() == '[') stk.Pop();
				else if(character == ']' && stk.Peek() != '[') return false;

				if(character == '}' && stk.Peek() == '{') stk.Pop();
				else if(character == '}' && stk.Peek() != '{') return false;
			}
		}

		return !Convert.ToBoolean(stk.Count);
	}
}

public class LeetCode
{
	public static void Main()
	{
		String str = "([}}])";
		Solution solution = new Solution();
		Console.WriteLine(solution.ValidParentheses(str));
	}
}

File: java/ValidParentheses.java
import java.util.*;

class Solution
{
	public boolean ValidParentheses(String s)
	{
		if(s.length() % 2 != 0) return false;

		Stack<Character> stk = new Stack<Character>();

		for(char c : s.toCharArray())
		{
			if(c == '(' || c == '[' || c == '{') stk.push(c);

			if(stk.size() == 0)
			{
				if(c == ')' || c == ']' || c == '}') return false;
			}

			if(stk.size() != 0)
			{
				if(c == ')' && stk.peek() == '(') stk.pop();
				else if(c == ')' && stk.peek() != '(') return false;

				if(c == ']' && stk.peek() == '[') stk.pop();
				else if(c == ']' && stk.peek() != '[') return false;

				if(c == '}' && stk.peek() == '{') stk.pop();
				else if(c == '}' && stk.peek() != '{') return false;
			}
		}

		return stk.size() == 0;
	}
}

public class Easy_Valid_Parentheses
{
	public static void main(String[] args)
	{
		String s = "([{}])";
		Solution solution = new Solution();
		System.out.println(solution.ValidParentheses(s));
	}
}
